flintrock.py

```diff
@@ -533,31 +528,82 @@ def launch_ec2(
         vpc_id=vpc_id,
         region=region)
 
+    num_instances = num_slaves + 1
     try:
         connection = boto.ec2.connect_to_region(region_name=region)
 
-        print("Launching {c} instances...".format(c=num_slaves + 1))
-
-        reservation = connection.run_instances(
-            image_id=ami,
-            min_count=(num_slaves + 1),
-            max_count=(num_slaves + 1),
-            key_name=key_name,
-            instance_type=instance_type,
-            placement=availability_zone,
-            security_group_ids=[sg.id for sg in security_groups],
-            subnet_id=subnet_id,
-            placement_group=placement_group,
-            tenancy=tenancy,
-            ebs_optimized=ebs_optimized,
-            instance_initiated_shutdown_behavior=instance_initiated_shutdown_behavior)
-
-        time.sleep(10)  # AWS metadata eventual consistency tax.
+        if spot_price is not None:
+            print("Requesting {c} spot instances at a max price of {p}".format(c=num_instances, p=spot_price))
+
+            instance_reqs = connection.request_spot_instances(
+                price=spot_price,
+                image_id=ami,
+                count=(num_instances),
+                key_name=key_name,
+                instance_type=instance_type,
+                placement=availability_zone,
+                security_group_ids=[sg.id for sg in security_groups],
+                subnet_id=subnet_id,
+                placement_group=placement_group,
+                ebs_optimized=ebs_optimized)
+
+            spot_instance_request_ids = [sir.id for sir in instance_reqs]
+
+            try:
+                while True:
+                    time.sleep(30)
+
+                    reqs = connection.get_all_spot_instance_requests(spot_instance_request_ids)
+                    active_instance_ids = []
+
+                    for r in reqs:
+                        req_id = r.id
+                        if r.state == 'active':
+                            active_instance_ids.append(r.instance_id)
+                            spot_instance_request_ids.remove(req_id)
+
+                    if len(active_instance_ids) == (num_instances):
+                        print("All {num_instances} granted".format(num_instances=num_instances))
+                        reservation_resultset = connection.get_all_reservations(active_instance_ids)
+
+                        all_instances = []
+                        for reservation in reservation_resultset:
+                            all_instances += reservation.instances
+                        break
+                    else:
+                        print("{grant} of {req} instances granted, waiting longer".format(
+                            grant=len(active_instance_ids), req=num_instances))
+            except:
```

Comment:
We shouldn't have bare excepts like this.

```python
except Exception as e:
    print(e)
    print("Canceling spot instance requests...")  # added ellipsis here
    ...
```

This is a bit more user-friendly.